src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+      if (currentCanalName != null
+          && (selectedLandTerritories.size() < 2 || selectedWaterTerritories.size() < 2)) {
+        if (JOptionPane.YES_OPTION != JOptionPane.showOptionDialog(null, "Canal '" + currentCanalName
+            + "' is incomplete. A canal needs at least 2 land and 2 water territories.\rDo you want to continue to deselect the canal?",
+            "Canal incomplete",
+            JOptionPane.YES_NO_OPTION, JOptionPane.WARNING_MESSAGE, null, null, JOptionPane.NO_OPTION))
+          return;
+        MapXMLHelper.canalDefinitions.remove(currentCanalName);
+      }
+      currentCanalName = null;
+      if (!selectedLandTerritories.isEmpty() || !selectedWaterTerritories.isEmpty()) {
+        clearSelection();
+        SwingUtilities.invokeLater(new Runnable() {
+          public void run() {
+            imagePanel.repaint();
+          }
+        });
+      }
+      return;
+    }
+
+    final Point point = e.getPoint();
+    final String newTerrName = findTerritoryName(point, polygons);
+
+    if (newTerrName == null)
+      return;
+    Boolean newTerrIsWater = MapXMLHelper.territoryDefintions.get(newTerrName).get(DEFINITION.IS_WATER);
+    if (newTerrIsWater == null)
+      newTerrIsWater = false;
+
+    final LinkedHashSet<String> newTerrNeighborsDiffType = getNeighborsByType(newTerrName, !newTerrIsWater);
+
+    final ArrayList<String> terrCanals = new ArrayList<String>();
+    if (newTerrIsWater) {
+      for (final Entry<String, Tuple<SortedSet<String>, SortedSet<String>>> canalDef : MapXMLHelper.canalDefinitions
+          .entrySet()) {
+        if (canalDef.getValue().getFirst().contains(newTerrName))
+          terrCanals.add(canalDef.getKey());
+      }
+    } else {
+      for (final Entry<String, Tuple<SortedSet<String>, SortedSet<String>>> canalDef : MapXMLHelper.canalDefinitions
+          .entrySet()) {
+        if (canalDef.getValue().getSecond().contains(newTerrName))
+          terrCanals.add(canalDef.getKey());
+      }
+    }
+
+    if (currentCanalName == null) {
+      if (newTerrNeighborsDiffType.size() < 2) {
+        JOptionPane.showMessageDialog(null, "The selected " + (newTerrIsWater ? "water" : "land")
+            + " territory is connected to less than 2 " + (!newTerrIsWater ? "water" : "land")
+            + " territories!", "Input Error", JOptionPane.ERROR_MESSAGE);
+        return;
+      }
+      if (!terrCanals.isEmpty()) {
+        terrCanals.add(NEW_CANAL_OPTION);
+        currentCanalName = (String) JOptionPane.showInputDialog(null,
+            "Which canal should be selected for territory '" + newTerrName + "?", "Choose Canal",
+            JOptionPane.QUESTION_MESSAGE,
+            null, terrCanals.toArray(new String[terrCanals.size()]), // Array of choices
+            terrCanals.get(0)); // Initial choice
+      }
+      if (terrCanals.isEmpty() || NEW_CANAL_OPTION.equals(currentCanalName)) {
+        String suggestedCanalName;

```

Comment:
Mark this is final to guarantee that you'll initialize and use it. Otherwise make it an `Optiona<String>` pls